### ISSUE TITLE
Cleaner build process

### ISSUE THREAD

**Issue by Aerijo** [atom/atom#21063]
## Summary

The build process has no structure right now, it would be nice to add some structure to it.

## Motivation

I'm trying to build Atom from source on Arch Linux. It's broken somewhere, but the build is opaque (no progress indication when installing modules) and hard to follow in general. Working out what step it's up to is not easy.

I would like to rewrite it to be built from "Task" components. These Tasks would provide a uniform interface to a chunk of "work" (mainly the existing exported functions in `/lib`). This interface would include
- A task name to be printed when it starts
- A set of log methods (subtask for when it's doing different things in the task, plus info / warn / error)

Making everything conform to the Task interface would provide a more uniform output, as well as make tweaking the output easier (e.g., formatting and log levels).

## Describe alternatives you've considered

There are npm packages that do something like this, but the idea is to be simple enough to just include in the `scripts` folder. It's important that the solution works before bootstrapping. 

## Additional context

I've started working on a concept for it [here](https://github.com/Aerijo/atom/tree/clean-build). It defines a `Task` class that all other tasks inherit from.

I experimented with support for hierarchical & concurrent execution of tasks, as well as an event based life cycle (see the`finished` type methods), but these things increase the complexity and would make it a burden to implement and maintain. I think the current design is non intrusive, while still providing a good base for structuring the build.

That branch also bumps the dependencies in `script` and tries to reorganise the build script to only load modules when they are used. Hopefully I'll be able to start testing and making PR's for this once I get the build working.

**Comment by therobgirl** [comment]
Omg stop it 🤪

**Comment by therobgirl** [comment]
##I love ATOM @
